### PR TITLE
Restructures parsing logic for performance

### PR DESCRIPTION
**Pull request by purew**
## Intro 

This change removes the translation to `Vec<char>` inside of the loop which makes a massive difference in decoding performance.

Further, I modularized the code, similar to a [python-decoding library does it](https://github.com/frederickjansen/polyline/blob/77893079ea38bfea3dc03135f94e2a27efffafbd/polyline/codec.py#L27-L52)

I added a new criterion benchmark which shows a 78% performance increase for the given small polyline,
and several orders of magnitude for a larger polyline.

## Benchmark of decoding of a small polyline 
```
bench polyline6 decoding
                        time:   [33.483 ns 33.527 ns 33.568 ns]
                        change: [-78.983% -78.787% -78.592%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild
```
## Benchmark of decoding of a large polyline 
```
bench HUGE polyline6 decoding
                        time:   [20.007 us 20.012 us 20.018 us]
                        change: [-99.975% -99.974% -99.974%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
```

The large polyline decoding went from 78milliseconds to 20 _micro-seconds_!

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -109,60 +109,21 @@ where
...
-            byte = at_index as u64 - 63;
-            index += 1;
-            result |= (byte & 0x1f) << shift;
-            shift += 5;
-            if byte < 0x20 {
-                break;
-            }
+    while index < chars.len() {
+        let (latitude_change, new_index) = trans(&chars, index);
+        if index >= chars.len() {
```
Comment (kinkard) [inline src/lib.rs:122]:
```suggestion
        if new_index >= chars.len() {
```
I see no point to check `index` in this line, maybe `new_index` should be checked instead?

**Comment by urschrei** [comment]
Hi @purew, sorry I didn't see this until now! We have two open decoding perf-related PRs, and there doesn't seem to be much overlap, but we'll figure out a way forward ASAP.

**Comment by urschrei** [comment]
This PR has a 30 % improvement over https://github.com/georust/polyline/pull/26, so going to merge this and release.

File: Cargo.toml [hunk 2]
```diff
@@ -1,8 +1,8 @@
 [package]
 name = "polyline"
 description = "Encoder and decoder for the Google Encoded Polyline format"
-version = "0.9.0"
-authors = ["Tom MacWright <tom@macwright.org>", "The GeoRust Developers"]
+version = "0.10.0"
```
Comment (michaelkirk) [inline Cargo.toml:4]: This change doesn't break any api's - so there's no reason for a semver breaking change, right?

It'd be nice if existing users could just get this perf bonus "for free" without having to bump their deps.

Comment (urschrei) [reply Cargo.toml:4]: Yes we could publish 0.9.1…

Comment (michaelkirk) [reply Cargo.toml:4]: I'll follow up with a PR.

Comment (michaelkirk) [reply Cargo.toml:4]: haha well, I was wrong, it *is* a breaking release after all: See #31

**Pull request merged**